### PR TITLE
WIP: first take on SB-458

### PR DESCRIPTION
**Pull request by inoxx03**
@cmoulliard @metacosm @geoand Still a WIP, but please let me know what you think. Thank you.

### CODE REVIEW

File: docs/topics/con_configmap-availability.adoc [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+java -jar {app-name}.jar -Dspring.profiles.active=developer
+----
+
+// make this a prerequisite:
+**Notes:**
+* You must assign view permissions to your OpenShift service account to access ConfigMap on OpenShift the service account needs at least view permissions i.e.:
++
+[source,bash,options="nowrap"]
+----
+oc policy add-role-to-user view system:serviceaccount:$(oc project -q):default -n $(oc project -q)
```
Comment (Ladicek) [inline docs/topics/con_configmap-availability.adoc]: Recommended practice, I think, is to have a dedicated service account for each application, not use the `default` one.

File: docs/topics/templates/document-attributes.adoc [hunk 2]
```diff
@@ -31,7 +31,7 @@
 :Thorntail: WildFly Swarm
 :SpringBoot: Spring Boot
 :VertX: Eclipse Vert.x
-:Node: Node.js
+:NodeJS: Node.js
```
Comment (Ladicek) [inline docs/topics/templates/document-attributes.adoc]: This was changed recently on a lot of places. I don't think this change is correct.

**Comment by Ladicek** [comment]
CC @liborfuka, could you please check? I don't feel very confident in the Spring Boot specifics.

File: docs/topics/con_configmap-availability.adoc [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+
+//openshift docs on configmaps
+//https://docs.openshift.org/latest/dev_guide/configmaps.html
+
+
+//replace Kubernetes with openshift?
+{deploymentPlatform} provides a resource named link:http://kubernetes.io/docs/user-guide/configmap/[`ConfigMap`] to externalize the
+parameters to pass to your application in the form of key-value pairs or embedded `application.properties|yaml` files.
+Your application automatically redeploys when the ConfigMap changes.
```
Comment (metacosm) [inline docs/topics/con_configmap-availability.adoc]: Is this true? Can you point to the reference stating this fact?

File: docs/topics/templates/document-attributes.adoc (hunk 2 above)
Comment (metacosm) [reply docs/topics/templates/document-attributes.adoc]: Also, I don't think that this change is appropriate within the scope of this PR.

File: docs/topics/con_configmap-availability.adoc [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+`ConfigMapPropertySource` searches the deployment configuration of your application for a Kubernetes `ConfigMap` object in which the `metadata.name` property matches the name of your {SpringBoot} application (set by the `spring.application.name` property).
```
Comment (metacosm) [inline docs/topics/con_configmap-availability.adoc]: I feel we're missing some context here but maybe it's because I'm not seeing the complete rendered doc? We should explain that this comes from Spring Cloud Kubernetes and point to the doc?

File: docs/topics/con_configmap-availability.adoc [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+Alternatively, you can specify a custom name for your ConfigMap.
+Set the value of the `spring.cloud.kubernetes.config.name` property in the `bootstrap.properties` file of your application project to the name of your ConfigMap.
+This enables the externalized configuration to be recognized and processed by your application.
+
+//(athese are alternatives, not a sequence)
```
Comment (metacosm) [inline docs/topics/con_configmap-availability.adoc]: Weird.

**Comment by metacosm** [comment]
Is there a way to check the rendered change in context? If yes, how?

**Comment by inoxx03** [comment]
@metacosm Please see the [rendered version](http://file.brq.redhat.com/~ssitani/spring-boot-master-2018-07-09.html#_managing_the_configuration_of_your_spring_boot_application).
I have made some updates to the explanatory paragraphs and temporarily placed the section into the Spring Boot Dev guide.

File: docs/topics/con_configmap-availability.adoc [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+:deploymentPlatform: Kubernetes
+
+//openshift docs on configmaps
+//https://docs.openshift.org/latest/dev_guide/configmaps.html
+//https://goo.gl/cQTA2B
+//https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#create-configmaps-from-files
+
+{deploymentPlatform} provides a resource named link:http://kubernetes.io/docs/user-guide/configmap/[`ConfigMap`] that allows you to externalize the configuration parameters that you pass to your application in the form of key-value pairs or embedded `application.properties` or `application.yaml` templates.
+The `ConfigMapPropertySource` provided by Spring Cloud Kubernetes searches the deployment configuration of your application for a ConfigMap object in which the `metadata.name` property matches the name of your {SpringBoot} application and applies the values to the deployment configuration of your application.
```
Comment (cmoulliard) [inline docs/topics/con_configmap-availability.adoc]: That should be great to showcase an example here about this concept introduced

File: docs/topics/con_configmap-availability.adoc [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+* Specify values of individual configuration properties directly in key-value format.
+{deploymentPlatform} processes the ConfigMap and maps the values to your application.
+This structure is suitable for updating specific properties within an otherwise static configuration.
++
+[source,yaml,options="nowrap"]
+----
+kind: ConfigMap
+apiVersion: v1
+metadata:
+name: demo
```
Comment (cmoulliard) [inline docs/topics/con_configmap-availability.adoc]: indentation is missing

File: docs/topics/con_configmap-availability.adoc [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+:deploymentPlatform: Kubernetes
+
+//openshift docs on configmaps
+//https://docs.openshift.org/latest/dev_guide/configmaps.html
+//https://goo.gl/cQTA2B
+//https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#create-configmaps-from-files
+
+link:http://kubernetes.io/docs/user-guide/configmap/[`ConfigMap`] is a {deploymentPlatform} resource type that allows you to externalize the configuration parameters that you pass to your application in the form of key-value pairs or embedded `application.properties` or `application.yaml` templates.
+The `ConfigMapPropertySource` provided by Spring Cloud Kubernetes searches the deployment configuration of your application for a ConfigMap object in which the `metadata.name` property matches the name of your {SpringBoot} application and applies the values to the deployment configuration of your application.
```
Comment (Ladicek) [inline docs/topics/con_configmap-availability.adoc]: "Deployment configuration" is also a name of an OpenShift resource, unrelated to this text. Hence it might be confusing. I'd reword to something like:

> The `ConfigMapPropertySource` provided by Spring Cloud Kubernetes looks for a ConfigMap object whose `metadata.name` property matches the name of your {SpringBoot} application and applies the values from the ConfigMap to your application.

File: docs/topics/con_configmap-availability.adoc [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+Without these permissions, your application will not be able to access the ConfigMap.
+
+[source,bash,options="nowrap",subs="attributes+"]
+--
+oc policy add-role-to-user view system:serviceaccount:$(oc project -q):{app-name} -n $(oc project -q)
+--
+====
+
+* Specify values of individual configuration properties directly in key-value format.
+{deploymentPlatform} processes the ConfigMap and maps the values to your application.
```
Comment (Ladicek) [inline docs/topics/con_configmap-availability.adoc]: This sentence doesn't make any sense.

File: docs/topics/con_configmap-availability.adoc [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: demo
+data:
+  pool.size.core: 1
+  pool.size.max: 16
+----
+
+* If your configuration template contains a property named `application.yaml`, {deploymentPlatform} processes the configuration template and creates the ConfigMap.
```
Comment (Ladicek) [inline docs/topics/con_configmap-availability.adoc]: This sentence doesn't make any sense.

File: docs/topics/con_configmap-availability.adoc [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+On application startup, {deploymentPlatform} checks if the ConfigMap contains a key with one of the following names:
```
Comment (Ladicek) [inline docs/topics/con_configmap-availability.adoc]: No it doesn't. Kubernetes doesn't know anything about Spring Boot.

Comment (geoand) [reply docs/topics/con_configmap-availability.adoc]: Did you mean Spring Cloud Kubernetes @inoxx03?  Also applies to next sentence as well

File: docs/topics/con_configmap-availability.adoc [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+  name: demo
+data:
+  application.yaml: |-
+    pool:
+      size:
+        core: 1
+        max:16
+----
+
+// not sure if multiple profiles are supported
```
Comment (geoand) [inline docs/topics/con_configmap-availability.adoc]: They should be

**Comment by inoxx03** [comment]
@lthon @geoand Thank you for the feedback. I'll update the PR. Maybe removing the `WIP` was not such a good idea :P

File: docs/topics/con_configmap-availability.adoc [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+
+* `pool.size.core`
+* `pool.size.maximum`
+
+You can inject configuration settings into your application in different ways depending on how you structure the data within the ConfigMap:
+
+
+// make this a prerequisite:
+// already documented in the Configmap booster doc
+[NOTE]
```
Comment (tradej) [inline docs/topics/con_configmap-availability.adoc]: This is indeed a prerequisite.

File: docs/topics/con_configmap-availability.adoc [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+Ensure that you have view permissions assigned to the service account of your application.
+Without these permissions, your application will not be able to access the ConfigMap.
+
+[source,bash,options="nowrap",subs="attributes+"]
+--
+oc policy add-role-to-user view system:serviceaccount:$(oc project -q):{app-name} -n $(oc project -q)
+--
+====
+
+* Specify values of individual configuration properties directly in key-value format.
```
Comment (tradej) [inline docs/topics/con_configmap-availability.adoc]: I would break out the list items into separate modules, either procedures or concepts, depending on how you write them. The current state, however, is not compatible with the modular guidelines.

**Comment by inoxx03** [comment]
This content has been moved to: http://snowdrop.me/guides/externalized-configuration
This PR will not be merged, closing.
